File: examples/fp-verification-demo-java/src/p2_counter_evidence.java
// P2 — Counter-Evidence: 反证搜寻，应被抑制
import java.util.concurrent.locks.*;
import java.util.concurrent.atomic.*;

class ResourceHandle implements AutoCloseable {
    private byte[] buffer = new byte[1024];
    public void close() { buffer = null; }  // RAII
}

public class p2_counter_evidence {
    private final Lock mutex = new ReentrantLock();
    private final AtomicInteger counter = new AtomicInteger(0);
    private int sharedState = 0;
    private int fileHandle = 0;

    // P2-01: try-with-resources 自动释放 — 非泄漏
    public void safeResource() {
        try (ResourceHandle h = new ResourceHandle()) {
            // 使用资源，自动 close()
        }
    }

    // P2-02: Lock 保护共享状态 — 非竞争
    public void safeIncrement() {
        mutex.lock();
        try {
            sharedState++;
        } finally {
            mutex.unlock();
        }
    }

    // P2-03: AtomicInteger 线程安全 — 非竞争
    public int safeCounter() {
        return counter.incrementAndGet();
    }

    // P2-04: 同步方法 — 非竞争
    public synchronized void syncWrite(String data) {
        // 同步写入
    }
}

File: examples/python-vuln-demo-no-answers/src/p2_counter_evidence.py
# P2 — Counter-Evidence: 反证搜寻
import threading
import os


class ResourceHandle:
    """RAII 资源句柄"""
    def __enter__(self):
        self.buffer = bytearray(1024)
        return self
    def __exit__(self, *args):
        self.buffer = None  # 自动释放


class SafeCounter:
    def __init__(self):
        self._lock = threading.Lock()
        self._value = 0
        self._file = None

    def safe_increment(self):

        with self._lock:
            self._value += 1
            return self._value

    def safe_resource(self):

        with ResourceHandle() as h:
            _ = len(h.buffer)

    def safe_write(self, data):

        self._lock.acquire()
        try:
            # 写入操作
            pass
        finally:
            self._lock.release()
